Feat: Initial settings

csbizz opened the pull request:
최초 스키마, 목데이터

csbizz committed:
Feat: Initial settings

최초 스키마, 목데이터

kipid commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,64 @@
+  url      = env("DATABASE_URL")
+}
+
+model User {
+  id String @id @default(uuid())
+
+  investments Investment[]
+  company     Company?
+}
```
name, nickname, email 정도 추가하시는게 좋아보이네요.
그리고 나중에 URL 복잡해지는거 싫으니, nickname 을 @id 로 지정하시는 것도 한번 생각해 봐요.
createdAt 과 updatedAt 도...

csbizz replied:
uuid식 id 외에 unique 필드를 두고 그걸 id로 사용하는 것에 대해서는 재원님과 논의해보겠습니다. 일단은 uuid로 해두고 변경사항 있으면 공유하는 것으로 하겠습니다.
요구하신 필드들 추가했습니다.

kipid commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,64 @@
+  // TODO 1의 주체가 User인가 Company인가 정해야됨
+}
+
+model Company {
+  // TODO 가상 투자 금액 총계를 직접 저장할 것인가, 매번 계산할 것인가
+  id                         String   @id @default(uuid())
+  name                       String   @unique
+  description                String
+  category                   Category
+  cumulativeInvestmentAmount Int // NOTE 이름 바꿀것, 누적 투자 금액(실제)
+  revenue                    Int // NOTE 매출액
+  employees                  Int // NOTE 고용 인원수
+  selectedCount              Int // NOTE 선택된 횟수
+
+  investments Investment[]
+  owner       User         @relation(fields: [ownerId], references: [id])
+  ownerId     String       @unique
+}
```
Company id 도 나중에 URL 깔끔하게 하기 위해서라도 [\w\-\_]+ pattern 에 맞는 nickname 을 설정하시는 것도 좋아보입니다.

여기도 createdAt, updatedAt 추가.

Int max 값이 21_4748_3647 이라고 하는데, 21억 넘기는 기업도 있을테니,
누적 투자 금액 BigInt? 였나로 써주세요. 그리고 이름이 너무 긴거 같아요. 좀만 짧게 바꿔주세요. AccumulInvest 정도?
revenue 도 BigInt 로...

csbizz replied:
회사명을 id로 하는 것에 대해서도 재원님과 논의해보겠습니다.
요구하신 필드명, 필드타입에 대한 사항 적용했습니다.

kipid commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,64 @@
+  ownerId     String       @unique
+}
+
+model Investment {
+  id       String @id @default(uuid())
+  name     String
+  amount   Int
+  comment  String
+  password String
+
+  user      User    @relation(fields: [userId], references: [id])
+  userId    String // NOTE 투자자
+  company   Company @relation(fields: [companyId], references: [id])
+  companyId String // NOTE 투자대상 기업
+}
```
createdAt, updatedAt 이 빠졌네요.

@@index([userId, updatedAt(sort: Desc)])
@@index(updatedAt(sort: Desc))
추가해서 query 가 빠르게 동작하도록 해주는게 좋을거 같습니다.

csbizz replied:
필드 추가했습니다.

인덱싱에 대해 찾아봤는데 updatedAt 필드에 인덱싱을 적용하는 경우 성능적으로 이득보다 실이 많은 것 같습니다. userId값을 기준으로 인덱스를 걸어뒀고, 차후 id 형식이 바뀔 경우 그쪽으로 변경해두겠습니다.

kipid commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,64 @@
+  companyId String // NOTE 투자대상 기업
+}
+
+enum Category {
+  categoryA
+  categoryB
+  categoryC
+}
```
카테고리를 enum 타입으로 하면 나중에 새로운 분야가 생겼을때 유지보수가 힘들거 같아요. 그냥 String 으로 저장하는게 어떨까요?

csbizz replied:
String으로 저장해서 superstruct로 입력단에서 검증하는것 vs enum 타입으로 저장단에서 검증하는것의 차이네요. migration이 필요 없다는 점에서 전자가 유지보수에는 더 쉬울 것 같습니다. 적용해두겠습니다.

kipid commented on data/mock.js:
```diff
@@ -0,0 +1,165 @@
+const user = [
```
export const USER (대문자로)

kipid commented on data/mock.js:
```diff
@@ -0,0 +1,165 @@
+    id: 'efbea72e-0fd0-49b0-8ec5-bcf3eaf5e335',
+  },
+];
+const company = [
```
export const COMPANY 대문자로...

kipid commented on data/mock.js:
```diff
@@ -0,0 +1,165 @@
+    ownerId: 'efbea72e-0fd0-49b0-8ec5-bcf3eaf5e335',
+  },
+];
+const investment = [
```
export const INVESTMENT 대문자로..

kipid reviewed (commented):
댓글들 남겼습니다. 리뷰도 좀 공부가 필요하네요. 어떻게 작성해야 하는지 좀 헤맸네요.

csbizz committed:
Fix: Edit schema

스키마 내에 타임스태프 추가, company category를 string으로 수정, 누적
투자 금액 필드명 변경, 누적 투자 금액과 매출액 필드 BigInt로 수정,
investment에 인덱스 추가
mock data의 변수명 변경

csbizz committed:
Fix: Add export on mock.js

kipid commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,76 @@
+  ownerId     String       @unique
+}
+
+model Investment {
+  id       String @id @default(uuid())
+  name     String
+  amount   Int
+  comment  String
+  password String
+
+  createdAt DateTime @default(now())
+  updatedAt DateTime @updatedAt
+
+  user      User    @relation(fields: [userId], references: [id])
+  userId    String // NOTE 투자자
+  company   Company @relation(fields: [companyId], references: [id])
+  companyId String // NOTE 투자대상 기업
+
+  @@index([userId])
+}
```
@@index(createdAt: Desc) 로라도 작성하는게 좋아보입니다. updatedAt 이 자주 바뀌어서 indexing 에 연산 소모가 많다면 투자 시점 별로 목록을 조회할 일도 많아보이니...

@@index([userId, (createdAt: Desc)] 로도 조회 많이 할테니 추가해 주면 좋을거 같아요.

csbizz replied:
수정했습니다.

kipid commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,76 @@
+  // TODO 1의 주체가 User인가 Company인가 정해야됨
+}
+
+model Company {
+  // TODO 가상 투자 금액 총계를 직접 저장할 것인가, 매번 계산할 것인가
+  id            String @id @default(uuid())
+  name          String @unique
+  description   String
+  category      String // 회사 카테고리. 입력단에서 검증해야함.
+  AccumulInvest BigInt // NOTE 누적 투자 금액(실제)
+  revenue       BigInt // NOTE 매출액
+  employees     Int // NOTE 고용 인원수
+  selectedCount Int // NOTE 선택된 횟수
+
+  createdAt DateTime @default(now())
+  updatedAt DateTime @updatedAt
+
+  investments Investment[]
+  owner       User         @relation(fields: [ownerId], references: [id])
+  ownerId     String       @unique
+}
```
owner 가 두명 이상일 경우는 없을까요?
coworker 로라도 유저를 여러명 넣는게 필요해 보이긴 합니다.

csbizz replied:
동업자가 몇명인지 모르는 상황에서 필드로 추가하면 정규화가 불가능합니다. user-company 사이에 junction table을 하나 만들어서 처리해야할 것 같네요. 스키마에 임시로 추가해두겠습니다.

kipid replied:
workers     User[]
  workerIds   String[]

이렇게 추가해보니 에러 안뜨던데, 이렇게 하면 안좋은 점이 있나요?

kipid commented on prisma/schema.prisma:
```diff
@@ -0,0 +1,76 @@
+  company     Company?
+}
+
+model Comparison {
+  id String @id @default(uuid())
+
+  createdAt DateTime @default(now())
+  updatedAt DateTime @updatedAt
+
+  // TODO ownerCompany
+  // TODO compareCompany * 5(최대)
+  // TODO 1:N 관계
+  // TODO 1의 주체가 User인가 Company인가 정해야됨
+}
```
이건 뭔 데이터가 들어가야 할지 감이 안잡히네요. 모여서 토론 한번 해야 할거 같아요.

csbizz replied:
네 금요일이나 주말쯔음에 재원님이랑 먼저 얘기해보고, 틀이 갖춰진 뒤에 본격적으로 작업하기 전 프론트-백 모두 모여서 한번 얘기하는거로 하죠

galaxy-78 replied:
저도 다 확인해보았습니다. 잘 모르겠는 부분도 있는데 백엔드 내일 논의해봅시다.

kipid reviewed (commented):
다시 코멘트 남겼습니다.

csbizz committed:
Fix: Add index on investment, Add idea about coworker
